pydantic/types.py

```diff
@@ -429,12 +430,17 @@ def validate(cls, value: Union[str]) -> Union[str]:
         if cls.curtail_length and len(value) > cls.curtail_length:
             value = value[: cls.curtail_length]
 
-        if cls.regex:
-            if not cls.regex.match(value):
-                raise errors.StrRegexError(pattern=cls.regex.pattern)
+        regex = cls._regex()
+        if regex:
+            if not regex.match(value):
```

Comment:
I'm afraid this might be slow since the compiled regex might not be cached.

Better to do the following which I think should work in both cases 

```suggestion
        if cls.regex:
            if not re.match(cls.regex, value):
```